### PR TITLE
Add datatype sniff button to edit dataset attributes page

### PR DESCRIPTION
**Pull request by qiagu**

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/controllers/dataset.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
     web
 )
 from galaxy.datatypes.display_applications.util import decode_dataset_user, encode_dataset_user
+from galaxy.datatypes import sniff
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/controllers/dataset.py:20]: This needs to be moved one line above, see our import order style: https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_smarkets.py

File: lib/galaxy/webapps/galaxy/controllers/dataset.py [hunk 2]
```diff
@@ -459,6 +461,24 @@ def __ok_to_edit_metadata(dataset_id):
...
+                    is_binary = check_binary(path)
+                    datatype = sniff.guess_ext(path, trans.app.datatypes_registry.sniff_order, is_binary=is_binary)
+                    trans.app.datatypes_registry.change_datatype(data, datatype)
+                    trans.sa_session.flush()
+                    trans.app.datatypes_registry.set_external_metadata_tool.tool_action.execute(
+                        trans.app.datatypes_registry.set_external_metadata_tool, trans, incoming={'input1': data},
+                        overwrite=False)  # overwrite is False as per existing behavior
+                    message = 'Sniffed and changed the type to %s.' % datatype
+            else:
+                return self.message_exception(trans, 'Changing datatype %s is not allowed.' % (data.extension))
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/controllers/dataset.py]:
```suggestion
                return self.message_exception(trans, 'Changing datatype "%s" is not allowed.' % (data.extension))
```

Comment (qiagu) [reply lib/galaxy/webapps/galaxy/controllers/dataset.py]: @nsoranzo Thanks

File: client/galaxy/scripts/mvc/dataset/dataset-edit-attributes.js [hunk 3]
```diff
@@ -172,6 +172,14 @@ var View = Backbone.View.extend({
                     onclick: function() {
                         self._submit("datatype", form);
                     }
+                }),
+                submit_datatype_sniff: new Ui.ButtonIcon({
+                    tooltip: _l("Sniff the datatype and change it."),
+                    title: _l("Sniff datatype"),
```
Comment (martenson) [inline client/galaxy/scripts/mvc/dataset/dataset-edit-attributes.js]: Can we go with something like 'Detect datatype' instead? Sniff could be confusing.

Comment (qiagu) [reply client/galaxy/scripts/mvc/dataset/dataset-edit-attributes.js]: Absolutely. Changed every 'sniff' to 'detect'. Thanks.

**Comment by qiagu** [comment]
@galaxybot test this

**Comment by jennaj** [comment]
Passes tests but presented a display problem for me in a default local inside a virtualenv. 

I'm not sure why. Does python need to be updated so the virtualenv is created correctly?

This is how the display problem presented: https://github.com/galaxyproject/usegalaxy-playbook/issues/46#issuecomment-439551830

**Comment by jennaj** [comment]
Passes tests in a local using this branch. Details: https://github.com/galaxyproject/usegalaxy-playbook/issues/46#issuecomment-443334407

This we should merge

**Comment by qiagu** [comment]
@jennaj Thanks!

**Comment by martenson** [comment]
@qiagu looks very good, thanks!

**Comment by qiagu** [comment]
@martenson  Thanks!

**Pull request merged**